src/net_processing.cpp

```diff
@@ -2058,6 +2058,8 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                 State(pfrom->GetId())->fProvidesHeaderAndIDs = true;
                 State(pfrom->GetId())->fWantsCmpctWitness = nCMPCTBLOCKVersion == 2;
             }
+            // ignore HB requests from peers without witness support
+            fAnnounceUsingCMPCTBLOCK = fAnnounceUsingCMPCTBLOCK && (pfrom->GetLocalServices() & NODE_WITNESS);
```

Comment:
Isn't `GetLocalServices()` always equal to what our services are, rather than our peers?  (I find this very confusing, but tracing through the code I can't see how this relates to what our peer sends us.)

I think this code should be checking against the fWantsCmpctWitness variable for the peer, not GetLocalServices().

Reply:
Indeed, it's gated on nCMPCTBLOCKVersion on my branch now, but it isn't showing up here because the PR is closed.   What I was actually intending to match on was pfrom->nServices ... I'm not actually sure what will happen if a peer requests v2 compact blocks while not sending NODE_WITNESS

Reply:
I think (but haven't checked recently) that NODE_WITNESS is only used for finding outbound peers that we want to be connected to, and that we don't use it anywhere else in our p2p protocol logic (eg deciding how to serialize transactions/blocks and negotiating compact block versions).  So I suspect it would work just fine?